packages/x-scheduler/src/primitives/utils/event-utils.ts

```diff
@@ -79,3 +123,403 @@ export function getEventDays(
+
+  if (adapter.isBefore(candidateDay, seriesStartDay)) {
+    return false;
+  }
+
+  switch (rule.freq) {
+    case 'DAILY': {
+      const daysDiff = diffIn(adapter, candidateDay, seriesStartDay, 'days');
+      return daysDiff % interval === 0;
+    }
+
+    case 'WEEKLY': {
+      const seriesWeek = adapter.startOfWeek(seriesStartDay);
+      const dateWeek = adapter.startOfWeek(candidateDay);
+
+      // If no BYDAY is provided in a WEEKLY rule, default to the weekday of DTSTART.
+      const byDay = weeklyByDayCodes(rule.byDay, [
+        NUM_TO_BYDAY[adapter.getDayOfWeek(seriesStartDay)],
+      ]);
+
+      const dateDowCode = NUM_TO_BYDAY[adapter.getDayOfWeek(candidateDay)];
+      if (!byDay.includes(dateDowCode)) {
+        return false;
+      }
+
+      const weeksDiff = diffIn(adapter, dateWeek, seriesWeek, 'weeks');
+      return weeksDiff % interval === 0;
+    }
+
+    case 'MONTHLY': {
+      const seriesMonth = adapter.startOfMonth(seriesStartDay);
+      const dateMonth = adapter.startOfMonth(candidateDay);
+
+      const monthsDiff = diffIn(adapter, dateMonth, seriesMonth, 'months');
+      if (monthsDiff % interval !== 0) {
+        return false;
+      }
+
+      // TODO (#19128): Add support for monthly recurrence modes (BYDAY rules)
+      if (rule.byDay?.length) {
+        // Only "same day-of-month" mode is supported right now.
+        throw new Error(
+          'MONTHLY supports only exact same date recurrence (day of month of DTSTART).',
+        );
+      }
+
+      // If no BYMONTHDAY is provided in a MONTHLY rule, default to the day of month of DTSTART.
+      const byMonthDay = rule.byMonthDay?.length
+        ? rule.byMonthDay
+        : [adapter.getDate(seriesStartDay)];
+
+      return byMonthDay.includes(adapter.getDate(candidateDay));
+    }
+
+    case 'YEARLY': {
+      const seriesYear = adapter.startOfYear(seriesStartDay);
+      const dateYear = adapter.startOfYear(candidateDay);
+
+      // Only exact "same month + same day" recurrence is supported.
+      // Any use of BYMONTH, BYMONTHDAY, BYDAY, or multiple values is not allowed.
+      if (rule.byMonth?.length || rule.byMonthDay?.length || rule.byDay?.length) {
+        throw new Error('YEARLY supports only exact same date recurrence (month/day of DTSTART).');
+      }
+
+      const sameMonth = adapter.getMonth(candidateDay) === adapter.getMonth(seriesStartDay);
+      const sameDay = adapter.getDate(candidateDay) === adapter.getDate(seriesStartDay);
+      if (!sameMonth || !sameDay) {
+        return false;
+      }
+
+      const yearsDiff = diffIn(adapter, dateYear, seriesYear, 'years');
+      return yearsDiff % interval === 0;
+    }
+
+    default:
+      return false;
+  }
+}
+
+export function estimateOccurrencesUpTo(
+  adapter: Adapter,
+  rule: RRuleSpec,
+  seriesStart: SchedulerValidDate,
+  date: SchedulerValidDate,
+): number {
+  if (adapter.isBefore(adapter.startOfDay(date), adapter.startOfDay(seriesStart))) {
```

Comment:
```suggestion
  if (adapter.isBeforeDay(date, seriesStart)) {
```

This should be safer when we have timezone